research-software-directory/RSD-as-a-service
# 1287 remote rsd search (RSD v3)

# Search for software in multiple RSD instances

Closes #1287

Changes proposed in this pull request:
* rsd administator can add remote RSD instances to be used in global search and shown on the software overview page.
* The remote name is suggested in the following order:
   * If the remote RSD defined "remote_name" property at rsd_info endpoint this label is suggested
   * If the remote RSD has not defined remote_name, the hostname of RSD is suggested as label {@hostname}
   * Admin can always change/overwrite suggested label for the remote RSD
* rsd administrator can suggest remote name by defining it in the "RSD info" section. 
* `rsd_info` table/endpoint is created to store the info about rsd instance. Basic endpoint information and remote_name  property is inserted in the table. `remote_name` prop is used in aggregated_software_overview.
* all users can search and filter software in all defined RSD instances from the software overview page or global search. 
* the "remote" software opens in the new tab and the "local" software opens in the same tab. 
* remote software has "opens in new tab" icon at the top right of the card
* if at least one remote RSD is defined, the additional filter (RSD Host) is shown on the software overview page.
* global search includes remote software. RSD host label is shown under the software name
* upgraded all dependencies to latest version: MUI v5.16.4 and React v19. **React upgrade to v19 is commited as "BREAKING CHANGE" and will result in major version bump of RSD images to v3.0.0**
* upgraded deprecated test methods

How to test:
* `make start` to build and generate test data
* navigate to software overview page and confirm that no RSD hosts filter is present (no remotes are defined yet). 
* navigate to Remotes page (Admin section)
* add Helmholtz remote: https://helmholtz.software/
* add RSD Saas remote: https://research-software-directory.org/
* add RSD dev remote: https://research-software.dev/ . **RSD dev is updated to this PR version to return RSD remote name. This suggestion is shown in the form of the icon (see last image)**.
* try to add not existing RSD remote. The validation should raise error and prevent you from adding invalid remote.
* run scrapers manually or wait for scrapers to run. You can stop & restart RSD to see the logs `docker compose down & docker compose up`
* after scrapers run examine the software overview. It should include software from remote RSD's too.
* you can also filter all software, including remote software. The counts shown in the filters should reflect all software  including remote software.
* confirm that additional filter is shown on the software overview page and you can use filter.
* confirm that local software is open in the same tab and remote software in the new tab.
* navigate to highlights page and confirm that additional filter "RSD Hosts" is not shown on the page and that filtering works properly
* navigate to Rsd info page of admin section. Add remote_name to your liking. Confirm that name of your rsd is changed in the RSD Host filter

## Example rsd info section
![image](https://github.com/user-attachments/assets/ce1082eb-20e1-4905-864f-e66b3b5148e1)

## Example add remote RSD
![image](https://github.com/user-attachments/assets/3e6b727c-7fbd-49f1-a40a-f167f78a03b6)

## Example remotes overview
![image](https://github.com/user-attachments/assets/fbec5a11-d658-420a-a36e-623342f8c2d8)

## Example software overview cards (showing source, remote icon and the additional filter)
![image](https://github.com/user-attachments/assets/9723789d-3ab0-45a6-9f0b-f4e107c7bd14)

## Example software overview list (with source)
![image](https://github.com/user-attachments/assets/28adbad2-20bd-465d-b104-8fcb8698bb6c)

## Example suggested remote RSD name (based on info received from remote)
![image](https://github.com/user-attachments/assets/7805185e-5f17-428c-8a38-de2f429aae38)

## Example global search includes domain for the remotes
![image](https://github.com/user-attachments/assets/2d6b071e-836a-4f74-9e83-71eebee5e5f7)

PR Checklist:
